docs/tech-specs/basic-spec.md

```diff
@@ -6,6 +6,256 @@
+* The consumer will take care of out of order message being pushed
+
+The considerations are -
+
+* The broker will ensure at least once webhook event is delivered
+* Consumer should have access to all the messages it failed to receive due to error on its part
+* Consumer could access past messages for replaying purpose
+
+Key attributes deliberately out of consideration -
+
+* There will be no guarantee of message order
+* There is no consideration for Consumer errors caused by bugs
+
+## Key Concepts
+
+| Concept | Definition |
+| -- | -- |
+| Channel | A channel is the broadcast highway for messages. |
+| Producer | Producer is the system generating a message and broadcasting it to a channel |
+| Consumer | Consumer is the system registering itself to listen to messages broadcasted to a specific channel |
+| DLQ / Dead Letter Queue | A queue (collection) of messages that failed deliver to the consumer for non 2XX response from client |
+| Message | The payload Producer wants distributed across the Consumer within a Channel |
+| Message Delivery | Reaching the consumer with the message should be considered as delivery, in case of success it would be marked as consumed, else end up in dead letter queue |
+| Message Status | Message has 2 status - Acknowledged and Out-for-delivery |
+| Message Delivery Status | Message Delivery status is associated with a Message, Channel and Consumer combination and is a enumeration of - In-flight, Retry-Delivery, Retry-In-flight, Delivered, Dead |
+| Rational-delay | Time delta for fail-safe mechanism to kick in |
+| Delivery-Timeout | How long will we wait for message delivery to finish before we step in |
+
+## Life-cycle of a Message
+
+### High-level Flow
+
+* Broker receives a **Message** addressed to a **Channel** from a **Producer**
+* Broker retrieves the **Consumers** interested in broadcast of a **Channel**
+* Broker _delivers_ **Message** to each **Consumer** retrieved in previous step
+
+### Key milestones in that flow
+
+* Acknowledge to **Producer** of receipt of **Message**
+* Track _delivery status_ of the **Message** against each **Consumer**
+
+### Non-trivial scenarios
+
+* Broker crashed before sending the acknowledgement
+* Broker crashed after sending the acknowledgement but before attempting to deliver the messages
+* Broker crashed attempting to deliver to some/all consumers but succeeded with others (if any)
+
+### Design considerations
+
+Based on the flow, non-trivial scenarios we want to derive at the following design considerations -
+
+* Acknowledgement should be tied to being able to store Message **and** sending the ACK signal, but not dependent on Producer received the ACK signal.
+* There should be a fail-safe way to ensure Acknowledged messages are attempted to deliver in case synchronous triggering of delivery process fails
+* Consumer may not be available, so we should have a retry policy
```

Comment:
What does retry policy mean here? Consumer supposes to pull for messages, if consumer is not available, the messages should be in the queue until the message timeout, then discard or push to dead letter queue based on configuration. Doesn't retry suppose to be handled by the consumer, like Celery?

Reply:
I will add this to the `Yet another....` section as well; but the basic premise is - for a service I do not want to add one web api and **a worker**; the worker addition adds complexity and increases failure point in addition to adding different solution patterns around scalability, fault-tolerance and high availability. Furthermore, none of the products you mentioned takes advantage of HTTP/2 multiplexing hence it does fully reach the full capability in terms of throughput performance.